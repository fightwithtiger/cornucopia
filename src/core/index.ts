export * from './jsx'
export * from './transform'
