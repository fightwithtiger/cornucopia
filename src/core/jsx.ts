import traverse from '@babel/traverse'
import { flatten } from '../utils'

export function jsx2ClassNameTree(jsxElementNodes: any) {
  jsxElementNodes = Array.isArray(jsxElementNodes) ? jsxElementNodes : [jsxElementNodes]

  const classMap: any = { children: [] }
  for (let i = 0; i < jsxElementNodes.length; i++) {
    const jsxNode = jsxElementNodes[i]
    const root = {
      children: [],
    }
    const child = resolveJsxElement(jsxNode, root)
    if (child.key && child.className) {
      classMap.children.push(child)
    } else {
      classMap.children.push(...child.children)
    }
  }

  return classMap
}
function resolveJsxElement(jsxNode: any, head: any) {
  const { openingElement = {} } = jsxNode
  const attributes = openingElement.attributes || []
  const classNameAttr = attributes.find((item: any) => item.name.name === 'className')

  const children = jsxNode.children?.filter((child: any) => ['JSXFragment', 'JSXElement', 'JSXExpressionContainer'].includes(child.type)).reduce((pre: any, cur: any) => {
    pre.push(...walkJSXChild(cur))
    return pre
  }, [])

  if (classNameAttr) {
    const classList = extractClassNameFromExpression(classNameAttr.value.expression)
    if (classList.length > 0) {
      head = {
        className: classList[0].className,
        key: classList[0].key,
        classList,
        children: [],
      }
    }
  }

  for (const c of children) {
    const child = resolveJsxElement(c, head)
    if (child.className === head.className && child.key === head.key) {
      continue
    }
    if (!child.className && !child.key && child.children) {
      head.children.push(...child.children)
    } else {
      head.children.push(child)
    }
  }

  return head
}

function walkJSXChild(c: any): any {
  const children = []
  if (['JSXElement', 'JSXFragment'].includes(c.type)) {
    children.push(c)
  } else if (c.type === 'JSXExpressionContainer') {
    children.push(...walkJSXChild(c.expression || {}))
  } else if (c.type === 'LogicalExpression') {
    children.push(...walkJSXChild(c.right))
  } else if (c.type === 'ConditionalExpression') {
    const { consequent, alternate } = c
    children.push(...walkJSXChild(consequent), ...walkJSXChild(alternate))
  }
  return flatten(children)
}

function extractClassNameFromExpression(expression: any) {
  const { type } = expression
  const list: any = []
  if (type === 'MemberExpression') {
    const res = extractMemberExpression(expression)
    list.push(res)
  } else if (type === 'TemplateLiteral') {
    const { expressions } = expression
    for (const exp of expressions) {
      const res = extractClassNameFromExpression(exp)
      list.push(...flatten(res))
    }
  }
  return list
}

function extractMemberExpression(expression: any) {
  const { object, property } = expression
  return {
    key: object.name,
    className: property.name,
  }
}

export function getJSXElements(ast: any) {
  const jsxElementNodes: any = []
  traverse(ast, {
    ReturnStatement(path: any) {
      if (path.node.argument && ['JSXFragment', 'JSXElement'].includes(path.node.argument.type)) {
        jsxElementNodes.push(path.node.argument)
      }
    },
  })

  return jsxElementNodes
}
export function getImportDeclarations(ast: any) {
  return ast.program.body.filter((item: any) => item.type === 'ImportDeclaration')
}
