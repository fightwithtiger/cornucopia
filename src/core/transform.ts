export function genStyleText(obj: any, topKey?: string) {
  const ignoreTopKey = !topKey
  const { children, classList = [] } = obj
  const matched = classList.find((item: any) => item.key === topKey)
  const { className, key } = ignoreTopKey ? classList[0] : matched || {}

  const isSameKey = (key: string) => (ignoreTopKey || key === topKey)

  let peerStr = ''
  const peersInTopKey = classList.filter((item: any) => isSameKey(item.key) && item.className !== className)
  for (const peer of peersInTopKey) {
    peerStr += ` ${peer.className} {}`
  }
  let childStr = ''
  for (const child of children) {
    childStr += genStyleText(child, topKey)
  }
  return isSameKey(key) ? `.${className} {${childStr}} ${peerStr}` : childStr
}
