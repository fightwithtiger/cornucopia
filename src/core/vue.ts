export function vueTemplate2ClassNameTree(ast: any, head: any = { children: [] }) {
  const { attrsMap = {}, children = [] } = ast
  const classAttr = attrsMap.class || ''
  if (classAttr) {
    const classNames = classAttr.split(/\s/g)
    head = {
      className: classNames[0],
      classList: classNames.map((name: string) => ({ className: name })),
      children: [],
    }
  }

  if (children && children.length > 0) {
    for (const c of children) {
      const child = vueTemplate2ClassNameTree(c)
      if (child.className === head.className) {
        continue
      }
      if (!child.className && child.children) {
        head.children.push(...child.children)
      } else {
        head.children.push(child)
      }
    }
  }

  return head
}
