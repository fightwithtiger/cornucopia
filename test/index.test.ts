import { describe, expect, it } from 'vitest'
import { compile } from 'vue-template-compiler'
import { vueTemplate2ClassNameTree } from '../src/core/vue'

describe('should', () => {
  it('exported', () => {
    const code = `
    <div>
    aaaaa
      <div class="header"></div>
    </div>
    `
    const { ast } = compile(`<template>${code}</template>`)
    const classMap = vueTemplate2ClassNameTree(ast)
    expect({ ast, classMap }).toMatchSnapshot()
  })
})
