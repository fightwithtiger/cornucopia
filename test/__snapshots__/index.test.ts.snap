// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should > exported 1`] = `
{
  "ast": {
    "attrsList": [],
    "attrsMap": {},
    "children": [
      {
        "attrsList": [],
        "attrsMap": {},
        "children": [
          {
            "static": true,
            "text": "
    aaaaa
      ",
            "type": 3,
          },
          {
            "attrsList": [],
            "attrsMap": {
              "class": "header",
            },
            "children": [],
            "parent": [Circular],
            "plain": false,
            "pre": undefined,
            "rawAttrsMap": {},
            "static": true,
            "staticClass": "\\"header\\"",
            "tag": "div",
            "type": 1,
          },
        ],
        "parent": [Circular],
        "plain": true,
        "pre": undefined,
        "rawAttrsMap": {},
        "static": true,
        "staticInFor": false,
        "staticProcessed": true,
        "staticRoot": true,
        "tag": "div",
        "type": 1,
      },
    ],
    "parent": undefined,
    "plain": true,
    "rawAttrsMap": {},
    "slotScope": undefined,
    "static": false,
    "staticRoot": false,
    "tag": "template",
    "type": 1,
  },
  "classMap": {
    "children": [
      {
        "children": [],
        "classList": [
          {
            "className": "header",
          },
        ],
        "className": "header",
      },
    ],
  },
}
`;
